### PR TITLE
JDK-8294266: Add a way to pre-touch java thread stacks

### PR DESCRIPTION
**Pull request by tstuefe**
When doing performance- and footprint analysis, `AlwaysPreTouch` option is very handy for reducing noise. It would be good to have a similar option for pre-touching thread stacks. In addition to reducing noise, it can serve as worst-case test for thread costs, as well as a test for NMT regressions.

Patch adds a new diagnostic switch, `AlwaysPreTouchStacks`, as a companion switch to `AlwaysPreTouch`. Touching is super-simple using `alloca()`. Also, regression test.

Examples:

NMT, thread stacks, 10000 Threads, default:

```
-                    Thread (reserved=10332400KB, committed=331828KB)                                       
                            (thread #10021)                                                                 
                            (stack: reserved=10301560KB, committed=300988KB)                                
                            (malloc=19101KB #60755)                                                         
                            (arena=11739KB #20037)                                                          
```

NMT, thread stacks, 10000 Threads, +AlwaysPreTouchStacks:

```
-                    Thread (reserved=10332400KB, committed=10284360KB)                        
                            (thread #10021)                                                    
                            (stack: reserved=10301560KB, committed=10253520KB)                 
                            (malloc=19101KB #60755)                                            
                            (arena=11739KB #20037)                                             
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294266](https://bugs.openjdk.org/browse/JDK-8294266): Add a way to pre-touch java thread stacks

### Reviewers
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [9cacecbf](https://git.openjdk.org/jdk/pull/10403/files/9cacecbf295a8b5dfafe7a93377d9e9d1c72be4b)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [9cacecbf](https://git.openjdk.org/jdk/pull/10403/files/9cacecbf295a8b5dfafe7a93377d9e9d1c72be4b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/10403/head:pull/10403` \
`$ git checkout pull/10403`

Update a local copy of the PR: \
`$ git checkout pull/10403` \
`$ git pull https://git.openjdk.org/jdk.git pull/10403/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10403`

View PR using the GUI difftool: \
`$ git pr show -t 10403`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10403.diff">https://git.openjdk.org/jdk/pull/10403.diff</a>

</details>

### CODE REVIEW

**Comment by shipilev** [comment]
Drive-by comment: there is `os::pretouch_memory(void* start, void* end, size_t page_size)` ;)

**Comment by tstuefe** [comment]
> Drive-by comment: there is `os::pretouch_memory(void* start, void* end, size_t page_size)` ;)

Good point. Had to cast the volatile away though.

**Comment by kimbarrett** [comment]
> > Drive-by comment: there is `os::pretouch_memory(void* start, void* end, size_t page_size)` ;)
> 
> Good point. Had to cast the volatile away though.

Just happened to notice this going by.  Maybe the signature for pretouch_memory should have volatile qualifiers?

**Comment by robehn** [comment]
Hey, a questions, what you think about just pre-touching a part of the stack?

**Comment by tstuefe** [comment]
> Hey, a questions, what you think about just pre-touching a part of the stack?

Hmm, I guess that would work, but what usage case did you have in mind? Also, how would we specify it, in percent of thread stack size, or as an absolute value?

**Review by robehn — APPROVED** [review]
I have seen some workload where just one or two threads are using large stacks and the majority is not.
I.e. -Xss256m and similar, having a few hundreds of threads and pre-touching both stacks and heap you may run out of physical pages and start swapping. It would be useful is such case to just pre-touch the first meg or so on the stack.

I thinking -Xss256m -XX:PreTouchStackSize=1m

Maybe it is not as useful as in my head...
Also I don't know if a user may retrieve good information about stack size usage.

I'll let you be the arbitrator.

I'll approve it as is, if this what you think is best, thanks.

**Comment by tstuefe** [comment]
> I have seen some workload where just one or two threads are using large stacks and the majority is not. I.e. -Xss256m and similar, 

Yikes...

>having a few hundreds of threads and pre-touching both stacks and heap you may run out of physical pages and start swapping. It would be useful is such case to just pre-touch the first meg or so on the stack.
> 
> I thinking -Xss256m -XX:PreTouchStackSize=1m
> 
> Maybe it is not as useful as in my head... Also I don't know if a user may retrieve good information about stack size usage.
> 
> I'll let you be the arbitrator.
> 

I'll think about it. It may make sense considering your example.

**Comment by theRealAph** [comment]
> > Drive-by comment: there is `os::pretouch_memory(void* start, void* end, size_t page_size)` ;)
> 
> Good point. Had to cast the volatile away though.

Be careful with that. On some OSes, touching more than N pages below the currently lowest-mapped stack page will segfault. Therefore you must touch from the top down.

**Comment by shipilev** [comment]
> > > Drive-by comment: there is `os::pretouch_memory(void* start, void* end, size_t page_size)` ;)
> > 
> > 
> > Good point. Had to cast the volatile away though.
> 
> Be careful with that. On some OSes, touching more than N pages below the currently lowest-mapped stack page will segfault. Therefore you must touch from the top down.

See `os::map_stack_shadow_pages()` for this.

I think the pre-touching like this should be done as "extended" "preemptive" stack bang up to `stack_shadow_safe_limit`. `os::pretouch_memory` is fine for heap memory, but for thread stacks I think we need to hook into the usual stack overflow machinery.

**Comment by tstuefe** [comment]
> > Be careful with that. On some OSes, touching more than N pages below the currently lowest-mapped stack page will segfault. Therefore you must touch from the top down.
> 
> See `os::map_stack_shadow_pages()` for this.
> 
> I think the pre-touching like this should be done as "extended" "preemptive" stack bang up to `stack_shadow_safe_limit`. `os::pretouch_memory` is fine for heap memory, but for thread stacks I think we need to hook into the usual stack overflow machinery.

Oh sure, I can touch in reverse order. But I wonder if that is really needed. I used alloca() to reserve the stack space, and it should take care of OS-side stack banging for me. On Windows it does. So in theory it should be functionally equivalent to `os::map_stack_shadow_pages()`.

**Comment by tstuefe** [comment]
not yet bot

**Comment by tstuefe** [comment]
not yet bot

**Comment by tstuefe** [comment]
not yet bot

File: src/hotspot/share/runtime/javaThread.cpp [hunk 1]
```diff
@@ -2072,3 +2076,22 @@ void JavaThread::vm_exit_on_osthread_failure(JavaThread* thread) {
...
+
+void JavaThread::pretouch_stack() {
+  // Given an established java thread stack with usable area followed by
+  // shadow zone and reserved/yellow/red zone, pretouch the usable area ranging
+  // from the current frame down to the start of the shadow zone.
+  const address end = _stack_overflow_state.shadow_zone_safe_limit();
+  if (is_in_full_stack(end)) {
+    char* p1 = (char*) alloca(1);
+    address here = (address) &p1;
+    if (is_in_full_stack(here) && here > end) {
```
Comment (gerard-ziemski) [inline src/hotspot/share/runtime/javaThread.cpp:2088]: Shouldn't we complain to the user here if false?

File: src/hotspot/share/runtime/javaThread.cpp [hunk 2]
```diff
@@ -2072,3 +2076,22 @@ void JavaThread::vm_exit_on_osthread_failure(JavaThread* thread) {
                                   os::native_thread_creation_failed_msg());
   }
 }
+
+void JavaThread::pretouch_stack() {
+  // Given an established java thread stack with usable area followed by
+  // shadow zone and reserved/yellow/red zone, pretouch the usable area ranging
+  // from the current frame down to the start of the shadow zone.
+  const address end = _stack_overflow_state.shadow_zone_safe_limit();
+  if (is_in_full_stack(end)) {
```
Comment (gerard-ziemski) [inline src/hotspot/share/runtime/javaThread.cpp:2085]: Shouldn't we complain to the user here if false?

File: test/hotspot/jtreg/runtime/Thread/TestAlwaysPreTouchStacks.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        Thread t = new Thread(null,
+                () -> System.out.println("Alive: " + stackSize),
+                "Thread-" + stackSize, stackSize);
+        return t;
+    }
+
+    public static void main(String[] args) throws InterruptedException, IOException {
+
+        int[] stackSizes = {
+          1024 * 256, 1024 * 512, 1024 * 1024 * 3
```
Comment (gerard-ziemski) [inline test/hotspot/jtreg/runtime/Thread/TestAlwaysPreTouchStacks.java:54]: I'm curious, how did you decide on these values?

**Review by gerard-ziemski — APPROVED** [review]

**Comment by gerard-ziemski** [comment]
LGTM, just had a few questions...

**Comment by tstuefe** [comment]
@gerard-ziemski , @jdksjolen 

Thanks guys, I had almost forgotten about this one :-)

@jdksjolen 

> To me alloca() seems like it should do the right thing. 

I think so too. 

> One question: Why is the global in gc_globals.hpp and not globals.hpp?

Its a companion switch to AlwaysPreTouch, which had been a GC switch for some reason. I think because its main intention was to pretouch heap.

File: src/hotspot/share/runtime/javaThread.cpp (hunk 2 above)
Comment (tstuefe) [reply src/hotspot/share/runtime/javaThread.cpp:2085]: Possibly, yes. Maybe even an assert.

File: src/hotspot/share/runtime/javaThread.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/runtime/javaThread.cpp:2088]: Same :-)

**Comment by tstuefe** [comment]
> FYI: We're currently in the process of investigating adding support for doing almost the exact opposite of this in JDK-8303215. In particular, we're looking to effectively "undo" the effects of huge page(s) temporarily having been used for the stack. There's potentially room for sharing some of the implementation aspects.

Interesting issue, but I don't think there is much overlap. Well, it depends on how that issue is solved. I'd like to get this issue out f the door though, the implementation is simple enough and can easily be changed later if we want to reuse code.

**Pull request closed without merging**